File: c/tree/validate-binary-search-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
bool isValidBST(struct TreeNode* root) {
    if(root == NULL)
        return true;

    struct TreeNode *node = root->left;
    while(node && node->right)
        node = node->right;

    if(node && node->val >= root->val)
        return false;

    node = root->right;
    while(node && node->left)
        node = node->left;

    if(node && node->val <= root->val)
        return false;

    return isValidBST(root->left) && isValidBST(root->right);
}

File: python/tree/validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isValidBST(self, root):
        if root is None:
            return True

        node = root.left
        while node and node.right:
            node = node.right

        if node and node.val >= root.val:
            return False

        node = root.right
        while node and node.left:
            node = node.left

        if node and node.val <= root.val:
            return False

        return self.isValidBST(root.left) and self.isValidBST(root.right)
